In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

Task 3.1.1

In [2]:
df = pd.read_csv("C:/Users/maksb/OneDrive/Documents/GitHub/2024-25b-fai1-adsai-MaksBurchard240894/Block B/diabetic_data_reg.csv", keep_default_na=False, na_values=["?"])

df.head()

C:\Users\maksb\AppData\Local\Temp\ipykernel_12448\1819032579.py:1: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("C:/Users/maksb/OneDrive/Documents/GitHub/2024-25b-fai1-adsai-MaksBurchard240894/Block B/diabetic_data_reg.csv", keep_default_na=False, na_values=["?"])


,encounter_id,patient_nbr,race,gender,age,weight,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,...,citoglipton,insulin,glyburide-metformin,glipizide-metformin,glimepiride-pioglitazone,metformin-rosiglitazone,metformin-pioglitazone,change,diabetesMed,readmitted
0,2278392,8222157,Caucasian,Female,[0-10),NaN,6,25,1,1,...,No,No,No,No,No,No,No,No,No,NO
1,149190,55629189,Caucasian,Female,[10-20),NaN,1,1,7,3,...,No,Up,No,No,No,No,No,Ch,Yes,>30
2,64410,86047875,AfricanAmerican,Female,[20-30),NaN,1,1,7,2,...,No,No,No,No,No,No,No,No,Yes,NO
3,500364,82442376,Caucasian,Male,[30-40),NaN,1,1,7,2,...,No,Up,No,No,No,No,No,Ch,Yes,NO
4,16680,42519267,Caucasian,Male,[40-50),NaN,1,1,7,1,...,No,Steady,No,No,No,No,No,Ch,Yes,NO


In [3]:
pd.DataFrame.isna(df)

,encounter_id,patient_nbr,race,gender,age,weight,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,...,citoglipton,insulin,glyburide-metformin,glipizide-metformin,glimepiride-pioglitazone,metformin-rosiglitazone,metformin-pioglitazone,change,diabetesMed,readmitted
0,False,False,False,False,False,True,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,True,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,True,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,True,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,True,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
101761,False,False,False,False,False,True,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
101762,False,False,False,False,False,True,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
101763,False,False,False,False,False,True,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
101764,False,False,False,False,False,True,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False


DataFrame size = 101766 x 50 (rows x columns)

In [4]:
df.replace("?", np.nan, inplace=True) #Credits: ChatGPT

pd.DataFrame.isna(df).sum()

encounter_id                    0
patient_nbr                     0
race                         2273
gender                          0
age                             0
weight                      98569
admission_type_id               0
discharge_disposition_id        0
admission_source_id             0
time_in_hospital                0
payer_code                  40256
medical_specialty           49949
num_lab_procedures              0
num_procedures                  0
num_medications                 0
number_outpatient               0
number_emergency                0
number_inpatient                0
diag_1                         21
diag_2                        358
diag_3                       1423
number_diagnoses                0
max_glu_serum                   0
A1Cresult                       0
metformin                       0
repaglinide                     0
nateglinide                     0
chlorpropamide                  0
glimepiride                     0
acetohexamide 

Missing Values per column:
race                         2273
..
weight                      98569
....
payer_code                  40256
medical_specialty           49949
......
diag_1                         21
diag_2                        358
diag_3                       1423

(others don't have missing values)

Evaluation: 

"Weight" has a large amounts of missing values and should be removed from the data.

"payer_code" and "medical_specialty" have around 50% of their data missing. I do not believe the values can easily be replaced so I believe both columns should be removed aswell.

"race", "diag_1", "diag_2", and "diag_3" have a low amount of missing data. I believe the rows that have the missing data should be removed.

In [5]:
df.dropna(axis=1, thresh=70000, inplace=True) #Cleans columns

df.dropna(inplace=True) #Cleans rows

pd.DataFrame.isna(df).sum()

encounter_id                0
patient_nbr                 0
race                        0
gender                      0
age                         0
admission_type_id           0
discharge_disposition_id    0
admission_source_id         0
time_in_hospital            0
num_lab_procedures          0
num_procedures              0
num_medications             0
number_outpatient           0
number_emergency            0
number_inpatient            0
diag_1                      0
diag_2                      0
diag_3                      0
number_diagnoses            0
max_glu_serum               0
A1Cresult                   0
metformin                   0
repaglinide                 0
nateglinide                 0
chlorpropamide              0
glimepiride                 0
acetohexamide               0
glipizide                   0
glyburide                   0
tolbutamide                 0
pioglitazone                0
rosiglitazone               0
acarbose                    0
miglitol  

In [74]:
df.dtypes

encounter_id                 int64
patient_nbr                  int64
race                        object
gender                      object
age                         object
admission_type_id            int64
discharge_disposition_id     int64
admission_source_id          int64
time_in_hospital             int64
num_lab_procedures           int64
num_procedures               int64
num_medications              int64
number_outpatient            int64
number_emergency             int64
number_inpatient             int64
diag_1                      object
diag_2                      object
diag_3                      object
number_diagnoses             int64
max_glu_serum               object
A1Cresult                   object
metformin                   object
repaglinide                 object
nateglinide                 object
chlorpropamide              object
glimepiride                 object
acetohexamide               object
glipizide                   object
glyburide           

All values have to be in a numeric value, so all "object" columns need to be changed.
Those are the following columns, along with explanations how to change them:
Below 3 columns have a small amount of distinct values, so they can be numbered.

race (5)
gender (3)
age (10)

Below 3 columns have numbers but are seen as "object" types. I believe that this is because of the non-whole numbers. So this needs to be a "float64" type.

diag_1
diag_2
diag_3

Below 2 columns have 4 values, of which half are different from the rest. Need to change those individually.

max_glu_serum
A1Cresult

Below 23 columns all have the same 4 values: "up", "down", "steady", "no". Can be mass changed into numerical values.

metformin
repaglinide
nateglinide
chlorpropamide
glimepiride
acetohexamide
glipizide
glyburide
tolbutamide
pioglitazone
rosiglitazone
acarbose
miglitol
troglitazone
tolazamide
examide
citoglipton
insulin
glyburide-metformin
glipizide-metformin
glimepiride-pioglitazone
metformin-rosiglitazone
metformin-pioglitazone

Below 2 columns have 2 values, boiling down to "yes" and "no".

change
diabetesMed

Below 1 column has 3 distinct values, which can be easily changed into numerical values.

readmitted

In [6]:
from sklearn.preprocessing import OneHotEncoder

##race and gender values should be one-hot encoded as they are not similar but age can just be simply labeled
##race
encoder = OneHotEncoder(sparse_output=False)
race_cat = df["race"]
encoded_array = encoder.fit_transform(df[["race"]])
encoded_df = pd.DataFrame (encoded_array, columns=encoder.get_feature_names_out(["race"]))
#result = pd.concat([df, encoded_df], axis=1)
#print(encoded_df)

##gender
gender_cat = df["gender"]
encoded_array_gender = encoder.fit_transform(df[["gender"]])
encoded_df_gender = pd.DataFrame (encoded_array_gender, columns=encoder.get_feature_names_out(["gender"]))
#print(encoded_df_gender)

##age
##all values: [0, 10) [10, 20) [20, 30) [30, 40) [40, 50) [50, 60) [60, 70) [70, 80) [80, 90) [90, 100)
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
le.fit(df[["age"]])
encoded_array_age = le.fit_transform(df[["age"]]) 
encoded_df_age = pd.DataFrame ({"Category_Age":encoded_array_age})
print(encoded_df_age)

#OneHotEncoder(df["race"])
#race_cat.head()


       Category_Age
0                 1
1                 2
2                 3
3                 4
4                 5
...             ...
98048             7
98049             8
98050             7
98051             8
98052             7

[98053 rows x 1 columns]


c:\Users\maksb\anaconda3\envs\BlockB\Lib\site-packages\sklearn\preprocessing\_label.py:97: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\maksb\anaconda3\envs\BlockB\Lib\site-packages\sklearn\preprocessing\_label.py:114: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [7]:
##max_glu_serum
le = LabelEncoder()
encoded_array_max_glu_serum = le.fit_transform(df[["max_glu_serum"]])
encoded_df_max_glu_serum = pd.DataFrame ({"Category max_glu_serum":encoded_array_max_glu_serum})
#(encoded_df_max_glu_serum)


##A1Cresult
le = LabelEncoder()
encoded_array_A1Cresult = le.fit_transform(df[["A1Cresult"]])
encoded_df_A1Cresult = pd.DataFrame ({"Category A1Cresult":encoded_array_A1Cresult})
print(encoded_df_A1Cresult)

       Category A1Cresult
0                       2
1                       2
2                       2
3                       2
4                       2
...                   ...
98048                   1
98049                   2
98050                   2
98051                   2
98052                   2

[98053 rows x 1 columns]


c:\Users\maksb\anaconda3\envs\BlockB\Lib\site-packages\sklearn\preprocessing\_label.py:114: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\maksb\anaconda3\envs\BlockB\Lib\site-packages\sklearn\preprocessing\_label.py:114: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [8]:
##The 23 long ones
le = LabelEncoder()
